art/attacks/evasion/shapeshifter.py

```diff
@@ -0,0 +1,1122 @@
+        if not self.graph_available:
+            self.graph_available = True
+
+            if self.texture_as_input:
+                # Check whether users provide a rendering function
+                rendering_function = kwargs.get("rendering_function")
+                if rendering_function is None:
+                    raise ValueError("Need a rendering function to use textures as input.")
+
+                # Build the TensorFlow graph
+                (
+                    self.project_texture_op,
+                    self.current_image_assign_to_input_image_op,
+                    self.accumulated_gradients_op,
+                    self.final_attack_optimization_op,
+                    self.current_variable,
+                    self.current_value
+                ) = self._build_graph(
+                    initial_shape=x.shape, custom_loss=custom_loss, rendering_function=rendering_function
+                )
+
+            else:
+                (
+                    self.project_texture_op,
+                    self.current_image_assign_to_input_image_op,
+                    self.accumulated_gradients_op,
+                    self.final_attack_optimization_op,
+                    self.current_variable,
+                    self.current_value
+                ) = self._build_graph(initial_shape=x.shape, custom_loss=custom_loss)
+
+        # Use mask or not
+        if self.texture_as_input:
+            # Check whether users have a mask
+            mask = kwargs.get("mask")
+            if mask is None:
+                mask = np.ones_like(x)
+        else:
+            mask = None
+
+        # Get victim class
+        victim_class = kwargs.get("victim_class")
+        if victim_class is None:
+            raise ValueError('Need to provide a victim class.')
+
+        if not isinstance(victim_class, int):
+            raise TypeError('Victim class must be of type `int`.')
+
+        # Get target class
+        target_class = kwargs.get("target_class")
+        if target_class is None:
+            logger.warning('Target class not provided, an untargeted attack is defaulted.')
+            target_class = victim_class
+
+        if not isinstance(target_class, int):
+            raise TypeError('Target class must be of type `int`.')
+
+        # Do attack
+        result = self._attack_training(
+            x=x,
+            y=label,
+            mask=mask,
+            target_class=target_class,
+            victim_class=victim_class,
+            project_texture_op=self.project_texture_op,
+            current_image_assign_to_input_image_op=self.current_image_assign_to_input_image_op,
+            accumulated_gradients_op=self.accumulated_gradients_op,
+            final_attack_optimization_op=self.final_attack_optimization_op,
+            current_variable=self.current_variable,
+            current_value=self.current_value
+        )
+
+        return result[1]
+
+    def _attack_training(
```

Comment:
This function seems to be used only once? Is it necessary?

Reply:
I suggest to let it like this. I tried to modulize the code to separate functions for easy maintenence and development. This function is the part of evaluation of the graph. So basically, I divide the code into 3 types: building graph, evaluating graph, and general code for the attack.